src/neoxp/Commands/BatchCommand.cs

```diff
@@ -122,6 +123,24 @@ internal async Task ExecuteAsync(IDirectoryInfo root, ReadOnlyMemory<string> com
                                 cmd.Model.Force).ConfigureAwait(false);
                             break;
                         }
+                    case CommandLineApplication<BatchFileCommands.Contract.Download> cmd:
+                        {
+                            if (cmd.Model.Height == 0)
+                            {
+                                throw new ArgumentException("Height cannot be 0. Please specify a height > 0");
+                            }
+                            
+                            var expressNode = chainManager.GetExpressNode();
+                            var result = await NodeUtility.DownloadParamsAsync(
```

Comment:
Might make sense to create a static helper method on `ContractCommand.Download` to do validation (single node chain, height > 0) and the calls to `DownloadParamsAsync` and `PersistContractAsync`. Not critical, but might be good to have validation all in one place

Reply:
Making sure I understand the request right; is the concern that the above 2 lines are duplicated in both `BatchCommand` and `ContractCommand.Download.ExecuteAsync()`? 
fwiw; only `BatchCommand` has the `height > 0` validation to ensure deterministic behaviour. `ContractCommand` on the other hand has `height = 0` as default where it means query the chain for the latest height.

Reply:
"Concern" is a bit of an overstatement. Introducing a method to just abstract away the calls to `DownloadParamsAsync` and `PersistContractAsync` isn't worth it. I was thinking that if you also validated `height > 0` and `consensus node count == 1`, it might become worth it. 

To do this, you'd need to lift the `stateAPI.GetStateHeightAsync` call out of `DownloadParamsAsync` and only call it in the direct call implementation (i.e. not from the batch command). However, I hadn't thought thru the ramifications of that change. Lifting `GetStateHeightAsync` out would then also force you to lift the `TransactionExecutor.TryParseRpcUri` call out since you need the URL to make `GetStateHeightAsync` call. At that point, we would have lost the value of the abstraction anyway.

Please do this:
* Add a check in `PersistContractAsync` implementation to ensure consensus node count is 1
* Also add a check before calling `DownloadParamsAsync`. No point of downloading the contract data only to throw when we go to persist
* Note, I left a separate comment that `DownloadParamsAsync` needs a better name

Reply:
I followed the request. Note that points 1 and 2 add quite some duplication, was that the intention? What we get now as flow is
- check node count
- call DownloadContractStateAsync
- call PersistContractAsync
    (internally)
   - check node count
   - call NodeUtility.PersistContract

Perhaps it's not worth spending too much time on as we'll quickly switch to multi-node support after this PR is merged and remove those checks again.

Reply:
It's a minor change, but I think there should be node count check in `NodeUtility.PersistContract` - utility methods like that need to make sure the operation they are about to make are allowed in the current context. Any other check we add would be for optimization - bailing out early before doing any work if the command is not allowed. 

* We should check node count as soon as possible after recognizing the user issued the contract download command. The checks that you added to BatchCommand and ContractCommand.Download are exactly what I wanted. 
* I can see the value in checking node count in `OnlineNode.PersistContractAsync`. Arguably better to bail out before making the RPC call
* The node check in `OfflineNode.PersistContractAsync` should simply be moved to `NodeUtility.PersistContract`. No need to check in both places.